Polkadot: Verify the workers at startup

naijauser opened the pull request:
# Description

Closes #8117

Today, `polkadot` checks that its PVF workers are running the same version at startup, but the workers' other security checks (landlock, seccomp, unshare/change-root, clone, etc.) only run when a worker is actually spawned. If the node is not yet part of the active validator set, an operator running a misconfigured or under-sandboxed worker has no way to find out — until the node starts validating, loses rewards, and degrades the network.

This PR runs those checks at startup too, the same way the version check already does: `polkadot` invokes `execute-worker --check-security-features --worker-dir-path <workers-path>` once before the node starts, and refuses to start if the worker binary is missing or fails its security checks. This way operators are informed immediately, rather than only once the node is selected as a validator.

## Integration

Node operators who pass `--workers-path` will now have their `execute-worker binary's security checks verified once at startup, before the node starts:

- If the worker binary is missing or cannot be spawned, the node now fails to start with `Error::ExecuteWorkerSpawnFailed` (binary path + underlying IO error), instead of panicking.
- If the worker reports failed security checks, the node fails to start with `Error::ExecuteWorkerFailedSecurityChecks` (binary path + worker dir path).

No changes to the normal worker spawn/handshake path (`polkadot-node-core-pvf`' `spawn_with_program_path`/`with_transient_socket_path`) — that continues to work exactly as before.

`polkadot-node-core-pvf-common::worker::check_security_status` is a new `pub fn`, used internally by the `decl_worker_main!` macro for the `--check-security-features` flag.

## Review Notes

- `polkadot/cli/src/command.rs`: new `verify_workers_security_checks` runs `<workers_path>/execute-worker --check-security-features --worker-dir-path <workers_path>` and maps a non-zero exit status / spawn failure to the new `Error` variants. It's called from `run()` right after CLI args are parsed, before any service/runner setup.
- `polkadot/node/core/pvf/common/src/worker/mod.rs`:
  - New `check_security_status(worker_kind, worker_dir_path) -> bool`: checks that the worker dir is readable and, on Linux, best-effort probes landlock/seccomp/change-root/clone — all logged as warnings and non-fatal, mirroring the best-effort philosophy already used for `SecurityStatus`.
  - `decl_worker_main!`'s argument parsing was reordered so `--worker-dir-path` is resolved (and the `do_security_checks` early exit can happen) before `--socket-path` is required, since the standalone check is never given a socket path.
- `polkadot/cli/src/error.rs`: added `ExecuteWorkerFailedSecurityChecks` and `ExecuteWorkerSpawnFailed` variants.
- Tests added:
  - `polkadot/cli/src/command.rs`: unit tests for `verify_workers_security_checks` covering success, worker-reported failure, missing binary, and the exact argument order, using stub shell-script "workers".
  - `polkadot/node/core/pvf/common/src/worker/mod.rs`: unit tests for `check_security_status` against existing/missing worker dirs (the "existing dir" case runs on a thread with a larger stack, since the Linux `clone(2)` probe needs more than the default `cargo test` thread stack).
  - `polkadot/node/core/pvf/tests/it/security_checks.rs`: new integration test file exercising `--check-security-features` against the real `prepare-worker`/`execute-worker` binaries.

# Checklist

* [x] My PR includes a detailed description as outlined in the "Description" and its two subsections above.
* [ ] My PR follows the [labeling requirements](
https://github.com/paritytech/polkadot-sdk/blob/master/docs/contributor/CONTRIBUTING.md#Process
) of this project (at minimum one label for `T` required)
    * External contributors: Use `/cmd label <label-name>` to add labels
    * Maintainers can also add labels manually
* [ ] I have made corresponding changes to the documentation (if applicable)
* [x] I have added tests that prove my fix is effective or that my feature works (if applicable)

naijauser committed:
verify workers at start up

naijauser committed:
Merge branch 'master' into naijauser/8117-verify-workers-at-startup
merge master

naijauser commented:
/cmd label T8-polkadot

naijauser committed:
add verify_on_start cli flag

naijauser committed:
add running_node_verify_on_start_fails test

naijauser committed:
add newline

naijauser committed:
add assertion to test

bkchr reviewed (changes requested):
Not really fixing the issue right now.

naijauser commented on polkadot/node/service/src/workers.rs:
```diff
@@ -97,6 +100,26 @@ pub fn determine_workers_paths(
+
+			worker_shutdown(worker_info, err);
+		}
+	}
```
Hello @bkchr, thanks for checking this out. I don't quite understand what you mean by `calling into the worker to do all the checks there.`

Could you clarify that bit?

naijauser replied:
I think I get it now. @bkchr 

Do you mean something like...

```rust
let output = Command::new(path)
        .arg("--check-can-enable-landlock")
        .arg("--check-can-enable-seccomp")
        .arg("--check-can-unshare-user-namespace-and-change-root")
        .arg("--check-can-do-secure-clone")
        .stderr(Stdio::piped())
        .stdout(Stdio::piped())
        .output()
        .with_context(|| format!("Failed to spawn worker {:?}", path))?;

    if !output.status.success() {
        worker_shutdown(worker_info, err);
    }
```

We do this inside `determine_workers_paths`.

naijauser committed:
remove earlier impl

naijauser committed:
add --check-all flag for worker and integrate check into determine_workers_paths

naijauser committed:
remove assert_cmd

naijauser committed:
add target_os

naijauser committed:
update error message

naijauser committed:
refactor macro

naijauser committed:
add and return error types

naijauser committed:
add separate tests for execute and prepare binaries

naijauser committed:
add commented out section

naijauser committed:
Merge branch 'master' into naijauser/8117-verify-workers-at-startup

naijauser committed:
Merge branch 'master' into naijauser/8117-verify-workers-at-startup

naijauser commented:
Hi @bkchr, bumping this up on your radar.

Pls take a look.

bkchr commented on polkadot/node/service/src/workers.rs:
```diff
@@ -97,6 +98,42 @@ pub fn determine_workers_paths(
+		.status()
+		.unwrap();
+
+	if exit_status.success() == false {
```
```suggestion
	if !exit_status.success() {
```

bkchr commented on polkadot/node/service/src/workers.rs:
```diff
@@ -97,6 +98,42 @@ pub fn determine_workers_paths(
+	let _ =  exec_worker_dir_path.pop();
+	let exit_status = Command::new(&exec_worker_path)
+	    // .arg("--socket-path")
+        // .arg(socket_path.as_ref().as_os_str())
```
```suggestion
```

bkchr commented on polkadot/node/core/pvf/common/src/worker/mod.rs:
```diff
@@ -81,7 +82,9 @@ macro_rules! decl_worker_main {
 					return
 				},
-
+				"--check-all" => {
```
Should be a better name

bkchr reviewed (commented):
Ty! Generally looking good, but is missing to call these checks at startup of the node + some other things need to be improved.

naijauser committed:
Update polkadot/node/service/src/workers.rs

Co-authored-by: Bastian Köcher <git@kchr.de>

naijauser committed:
Update polkadot/node/service/src/workers.rs

Co-authored-by: Bastian Köcher <git@kchr.de>

naijauser committed:
Update polkadot/node/service/src/workers.rs

Co-authored-by: Bastian Köcher <git@kchr.de>

naijauser committed:
Merge branch 'master' into naijauser/8117-verify-workers-at-startup

naijauser committed:
fmt

naijauser committed:
add security check on node start

naijauser committed:
add return

naijauser committed:
Merge branch 'master' into naijauser/8117-verify-workers-at-startup

github-actions[bot] committed:
Update from github-actions[bot] running command 'prdoc --audience node_dev --bump minor'

naijauser committed:
refactor logic

naijauser committed:
remove unused code

naijauser committed:
re-run build to update cargo lock

naijauser committed:
fmt

naijauser committed:
Merge branch 'master' into naijauser/8117-verify-workers-at-startup

github-actions[bot] committed:
Update from github-actions[bot] running command 'prdoc --audience node_dev --bump minor'

naijauser committed:
Merge branch 'master' into naijauser/8117-verify-workers-at-startup

naijauser committed:
update Cargo.lock

naijauser committed:
refactor the feature and add tests

naijauser committed:
update cargo

naijauser committed:
update test to use rusty-fork-test

naijauser committed:
remove rusty-fork

naijauser committed:
remove space

naijauser committed:
update prdoc

naijauser committed:
remove newline

sandreim committed:
Cumulus: introduce authority discovery for parachains (#11954)

Add authority discovery to parachains. Collators now run an
`authority_discovery` worker for the parachain network, publishing their
address to the parachain DHT and resolving each other from it.
Every parachain authority connects directly to every other authority,
forming a full collator-to-collator mesh on the block-announce protocol.

 ## Why

The mesh fixes a real problem on large parachain networks. Without it, a
freshly built block needs to go through multiple hops until it reaches
the next author. Each hop adds 3x the latency between the two peers
(block announce, block request, block response).
When blocks arrive too late, next authors are forced to build on stale
parents, and the chain forks, reducing block confidence and increasing
parachain block times.

**A followup PR will enable this for AH and PC.**

---------

Signed-off-by: Andrei Sandu <andrei-mihail@parity.io>
Co-authored-by: cmd[bot] <41898282+github-actions[bot]@users.noreply.github.com>
Co-authored-by: Bastian Köcher <git@kchr.de>

naijauser commented:
Hi @bkchr, I've made extensive updates to this PR.

Please take another look.

eskimor committed:
Move storage design to web3-storage repo (#12355)

The design should live along-side its implementation. A bit dumb, that I
realize this after the merge, but anyways. Here we go. The design was
already copied to the implementation repo and this is also where it
should belong. Removing it here.

https://github.com/paritytech/web3-storage/pull/170

alvicsam committed:
ci: fix zombienet logging (#12344)

cc https://github.com/paritytech/devops/issues/5376

---------

Co-authored-by: Javier Viola <javier@parity.io>

alexggh committed:
statement-store: log account when statements get rejected (#12357)

... to make debugging easier

Signed-off-by: Alexandru Gheorghe <alexandru.gheorghe@parity.io>

skunert committed:
Bump litep2p to 0.14.2 (#12358)

Bump litep2p to get latest [bitswap
fixes](https://github.com/paritytech/litep2p/releases/tag/v0.14.2).

tdimitrov committed:
Logging improvements for the collator revamp (#12282)

Some logging updates:

- Decrease log levels in `wait_for_first_leaf` to DEBUG, to avoid
startup spam
- Use `warn_if_frequent` for fetch errors
- Log assignment changes on view change
- pick_best_advertisement: trace logs for each outcome
- update_view: log scheduling parent <-> assigned core mapping
- update_view: log sp removal
- handle_seconded_collation: logs for each error case
- PeerManager: log reputation updates

Partially addresses
https://github.com/paritytech/polkadot-sdk/issues/10402

---------

Co-authored-by: cmd[bot] <41898282+github-actions[bot]@users.noreply.github.com>

sigurpol committed:
nomination-pools: snapshot rewards before set_commission_max lowers current commission (#12397)

`set_commission_max` force-lowers `commission.current` (via
`try_update_max`) when the new max is below the active rate, but did not
first call `update_records`. Rewards accrued at the higher rate since
the last snapshot were therefore re-rated at the new lower rate on the
next `update_records`, crediting the differential `(old_current -
new_max) * accrued` to members instead of the commission payee.

The fix snapshots the reward pool at the current commission before the
cut, mirroring the ordering already used in `set_commission`.

michalkucharczyk committed:
fatxpool: preserve order of same-priority transactions on re-orgs (#12362)

This PR improves handling the order of transactions during re-orgs.
Previously transactions from retracted blocks were resubmitted in
reverse order (old best → common ancestor); now they are resubmitted in
execution order (common ancestor → old best), so equal-priority
transactions keep their original order via the pool's FIFO tie-break.

The original order is not always preserved; this is a best-effort
approach (not a guarantee) that works well for most typical forks,
preserving order where possible:

#### Works — equal priority (e.g. same call, no tips)

```
retracted:           B1[TA0]  B2[TB0]  B3[TA1]  B4[TB1]
resubmitted order:   B'x[TA0, TB0, TA1, TB1]
```

Both per-account nonce order **and** the cross-account interleaving are
preserved.

#### Doesn't work — different priority (tips / different weights)

Priority wins over original order. With `TA1` boosted (lower weight or
higher tip):

```
retracted:           B1[TA0]  B2[TB0]  B3[TA1]  B4[TB1]
resubmitted order:   B'x[TA0, TA1, TB0, TB1]
```

Per-account nonce order is still safe (`TA1` never beats `TA0`), but the
cross-account order follows priority, not the blocks.

#### Note: Where does priority come from?
For a signed transaction, priority is set by `ChargeTransactionPayment`
based on the tip per unit of the scarcer resource (weight or length):
```
priority = (tip + 1) * max_tx_per_block      
```
For more details refer to
[`get_priority`](https://github.com/paritytech/polkadot-sdk/blob/968eaa447002e599ee1c9deac6d41eca230204ae/substrate/frame/transaction-payment/src/lib.rs#L871)

---------

Co-authored-by: cmd[bot] <41898282+github-actions[bot]@users.noreply.github.com>

cirko33 committed:
Decouple unbonding pool bound from bonding duration (#12323)

The with_era map is now bounded by the fixed MaxUnbondingPools, and the
merge cutoff uses an effective window of MaxUnbondingPools -
bonding_duration. This prevents a lowered bonding duration from
shrinking the bound and making stored state undecodable.

---------

Co-authored-by: cmd[bot] <41898282+github-actions[bot]@users.noreply.github.com>
Co-authored-by: Paolo La Camera <paolo@parity.io>
Co-authored-by: Ankan <ankan.anurag@gmail.com>

pepoviola committed:
changes to use zombie-bite (#12247)

In order to use [zombie-bite](https://github.com/paritytech/zombie-bite)
(to fork and spawn a live network with the _state_) with the same client
binaries (for testing releases), we need to allow to set this value
`DISPUTE_CANDIDATE_LIFETIME_AFTER_FINALIZATION` to __1__. I think the
best approach here is to use an _env var_, since we _only_ need it for
our use case. Cc @eskimor wdyt?

Thx!

---------

Co-authored-by: cmd[bot] <41898282+github-actions[bot]@users.noreply.github.com>

naijauser committed:
Merge branch 'master' into naijauser/8117-verify-workers-at-startup-working
merge master

naijauser committed:
fmt

naijauser committed:
fmt

naijauser commented:
/cmd prdoc --audience node_dev node_operator --bump major --force

github-actions[bot] committed:
Update from github-actions[bot] running command 'prdoc --audience node_dev node_operator --bump major --force'

naijauser committed:
Merge branch 'master' into naijauser/8117-verify-workers-at-startup
merge master

naijauser committed:
fix fmt

naijauser commented:
/cmd prdoc --audience node_dev node_operator --bump patch --force

github-actions[bot] committed:
Update from github-actions[bot] running command 'prdoc --audience node_dev node_operator --bump patch --force'

naijauser committed:
Merge branch 'naijauser/8117-verify-workers-at-startup' into naijauser/8117-verify-workers-at-startup-working
merge

naijauser committed:
fix semver

naijauser committed:
update test

naijauser committed:
Merge branch 'master' into naijauser/8117-verify-workers-at-startup

naijauser committed:
fmt

naijauser committed:
update prdoc

naijauser committed:
update

naijauser committed:
fix semver
